Fix restoring files ownership as non-root user

middelink opened the pull request:
Like tar, it should be possible to restore files and skip the setting ownership of files for non-root users.

Fixes #655

fd0 commented on src/restic/node.go:
```diff
@@ -160,32 +160,31 @@ func (node *Node) CreateAt(path string, repo Repository, idx *HardlinkIndex) err
 	}
 
+	// Do ownership last and only if you are acting as root.
+	if os.Getuid() == 0 {
```
That's not a good test, `EUID == 0` is sometimes sufficient, or having `CAP_CHOWN` available. I'd just try it last, and return the error (I have no better idea for it).

fd0 commented on src/restic/node.go:
```diff
@@ -160,32 +160,29 @@ func (node *Node) CreateAt(path string, repo Repository, idx *HardlinkIndex) err
 	}
 
+	// Change ownership last, as we lose permissions on the file.
+	if err = lchown(path, int(node.UID), int(node.GID)); err != nil {
```
For suid files (as in the test that fails), the suid bits are reset when the file is chowned. And the ctime is also changed. So I'd suggest moving `lchown()` to the beginning of the function again, but only return the error at the end.

middelink replied:
Yeah, live suxs, i'll look into recoding the recoded code... I got sidetracked a bit by transformation functions.

fd0 replied:
You don't need to apologize, thanks for your work!

Commit:
Fix restoring files as non-root user

As we cannot reliably detect in advance if we can set ownership, permissions,
timestamps or ext attributes, execute ALL the requested changes before
returning the first error we found.

Report total errors at end of restore and stop printing entire stacktraces
where just the error message is sufficient.

Fixes #655

middelink commented:
Done, this works locally to my personal satisfaction :)

fd0 commented:
Thanks!
